homeassistant/components/screenlogic/__init__.py

```diff
@@ -255,14 +250,20 @@ async def async_turn_off(self, **kwargs) -> None:
         """Send the OFF command."""
         await self._async_set_circuit(ON_OFF.OFF)
 
-    async def _async_set_circuit(self, circuit_value) -> None:
-        async with self.coordinator.api_lock:
-            success = await self.hass.async_add_executor_job(
-                self.gateway.set_circuit, self._data_key, circuit_value
+        # Turning off spa or pool circuit may require more time for the
+        # heater to reflect changes depending on the pool controller,
+        # so we schedule an extra refresh a bit farther out
+        if self._data_key in PRIMARY_CIRCUIT_IDS:
+            async_call_later(
+                self.hass, HEATER_COOLDOWN_DELAY, self._async_refresh_timed
             )
 
-        if success:
+    async def _async_set_circuit(self, circuit_value) -> None:
+        if await self.gateway.async_set_circuit(self._data_key, circuit_value):
             _LOGGER.debug("Turn %s %s", self._data_key, circuit_value)
+            await self.coordinator.async_refresh()
+            # Second debounced refresh to catch any secondary
+            # changes in the device
             await self.coordinator.async_request_refresh()
```

Comment:
Since this logic is in 4 places, it could be its own function in the base entity.

Reply:
https://github.com/home-assistant/core/pull/60466/commits/f2a27c7ea7ae5cd43b9ccd488588a036d35b3266